lib/devise/models/trackable.rb

```diff
@@ -31,8 +31,13 @@ def update_tracked_fields(request)
       end
 
       def update_tracked_fields!(request)
+        # We have to check if the user is already persisted before running
+        # `save` here because invalid users can be saved if we don't.
+        # See https://github.com/plataformatec/devise/issues/4673 for more details.
+        return if new_record?
+
         update_tracked_fields(request)
```

Comment:
Should not the `update_tracked_fields` still be called when it is a new record?

Reply:
When the signup succeeds the record would already be persisted here. This is being called from a Warden's `after_set_user` [callback](https://github.com/plataformatec/devise/blob/master/lib/devise/hooks/trackable.rb#L9). The callback is triggered when we call `set_user` from the `sign_in` [helper](https://github.com/plataformatec/devise/blob/master/lib/devise/controllers/sign_in_out.rb#L50), which is called after the record is created in [registrations controller](https://github.com/plataformatec/devise/blob/master/app/controllers/devise/registrations_controller.rb#L24).

Reply:
Makes sense. Is there any way we can test this to make sure it doesn't happen anymore?

Reply:
The test introduced in https://github.com/plataformatec/devise/pull/4674/files#diff-8aa58f219abef0280c8857f9161c087bR42 is still passing, so I think we're good to go.

Reply:
But how about the tests to make sure that every request don't run validations?

Reply:
I see, I'm not sure of how we should test this 🤔 
Do you think we should check here if the validations were run or in the sessions controller?

Reply:
One way to test this is write an integration test, add a validation to the model that sets a global value in the class if the validation is run. This way we can test if a request to any action sets this global value.

Reply:
I think it's worth to write a unit test to the `Trackable` module to check if the record is saved not running the validations 🤔 .